Handle resource update conflicts without dropping local changes

ELENAGER opened the pull request:
Problem: When we try to resolve a conflict during a resource update, we reload the object from the API server into the same variable. This replaces the in-memory object with the server version — including the old Spec/Status — and discards our changes.

Solution: Fetch a fresh copy of the object from the API server and merge our changes, preserving existing modifications.

Fixes: https://issues.redhat.com/browse/DFBUGS-4535

raaizik commented:
I would avoid just copying the ResourceVersion and retrying the update because that can overwrite concurrent changes and alter the expected behavior. I would instead use a standard retry on conflict pattern: get the latest object, merge only the changes you want, and then update. Alternatively for small field changes use a patch to ensure only your modifications are applied. I'd avoid copying ResourceVersion in this way - it is not safe and can lead to lost updates. IMHO the underlying PV reconciliation issue should be diagnosed and fixed without overwriting other fields.

ELENAGER commented:
> I would avoid just copying the ResourceVersion and retrying the update because that can overwrite concurrent changes and alter the expected behavior. I would instead use a standard retry on conflict pattern: get the latest object, merge only the changes you want, and then update. Alternatively for small field changes use a patch to ensure only your modifications are applied. I'd avoid copying ResourceVersion in this way - it is not safe and can lead to lost updates. IMHO the underlying PV reconciliation issue should be diagnosed and fixed without overwriting other fields.

@raaizik I agree, changed the code to merge changes. thanks for the comment!

abhijeet219 reviewed (approved):
lgtm
